### PR TITLE
Idp redesign

### PR DESCRIPTION
**Pull request by mtestrot**
Please review the refactoring.

### CODE REVIEW

File: idp/authmiddleware/authmiddleware.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
+// Package authmiddleware contains a http middleware for the authentication with the IdentityProvider-App
+package authmiddleware
+
+import (
```
Comment (weberaj) [inline idp/authmiddleware/authmiddleware.go]: Run goimports over imports. (Grouped imports by standard and external packages)

File: idp/authmiddleware/authmiddleware.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+//		mux := http.NewServeMux()
+//		mux.Handle("/hello", authenticate(helloHandler()))
+//		mux.Handle("/resource", authenticate(resourceHandler()))
+//		mux.Handle("/resource4ExternalUsers", authenticateExternal(resource4ExternalUsersHandler()))
+//	}
+//
+//	func helloHandler() http.Handler {
+//		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+//			// get user from context
+//			principal,_ := authmiddleware.PrincipalFromCtx(r.Context())
```
Comment (weberaj) [inline idp/authmiddleware/authmiddleware.go]: Maybe it's a good idea to handle and not ignore the error in this example?

File: idp/authmiddleware/authmiddleware.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+				}
+				return
+			}
+			systemBaseUri, gSBErr := getSystemBaseUriFromCtx(ctx)
+			if gSBErr != nil {
+				logError(ctx, fmt.Sprintf("error reading SystemBaseUri from context because: %v\n", gSBErr))
+				http.Error(rw, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+				return
+			}
+			tenantId, gTErr := getTenantIdFromCtx(req.Context())
```
Comment (weberaj) [inline idp/authmiddleware/authmiddleware.go]: You got already the request context, use ctx here.

File: idp/idpclient/client.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
+// name 'idpclient' is used although it repeats the name of the outer package 'idp'
```
Comment (weberaj) [inline idp/idpclient/client.go]: Do we need to explain this? Or should we instead explain in a short sentence what the idpclient package does?

File: idp/idpclient/client.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+// in order to improve readability functions like e.g. idpclient.New() or idpclient.HttpClient()
+package idpclient
+
+import (
```
Comment (weberaj) [inline idp/idpclient/client.go]: Run goimports over imports. (Grouped imports by standard and external packages)

File: idp/idpclient/client.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+		if validFor > 0 {
+			c.principalCache.Set(cacheKey, p, validFor)
+		}
+		return &p, nil
+	case http.StatusUnauthorized:
+		_, _ = ioutil.ReadAll(resp.Body) // client must read to EOF and close body cf. https://godoc.org/net/http#Client
+		return nil, nil
+	default:
+		responseMsg, _ := ioutil.ReadAll(resp.Body)
+		return nil, fmt.Errorf(fmt.Sprintf("Identityprovider '%s' returned HTTP-Statuscode '%d' and message '%s'",
```
Comment (weberaj) [inline idp/idpclient/client.go]: fmt.Sprintf is not needed, as you already use fmt.Errorf.

File: idp/idpclient/client_test.go [hunk 7]
```diff
@@ -0,0 +1,380 @@
+package idpclient_test
+
+import (
```
Comment (weberaj) [inline idp/idpclient/client_test.go]: Run goimports over imports. (Grouped imports by standard and external packages)

File: idp/authmiddleware/authmiddleware.go (hunk 1 above)
Comment (mtestrot) [reply idp/authmiddleware/authmiddleware.go]: Done

File: idp/authmiddleware/authmiddleware.go (hunk 2 above)
Comment (mtestrot) [reply idp/authmiddleware/authmiddleware.go]: Done.

File: idp/authmiddleware/authmiddleware.go (hunk 3 above)
Comment (mtestrot) [reply idp/authmiddleware/authmiddleware.go]: Done

File: idp/idpclient/client.go (hunk 5 above)
Comment (mtestrot) [reply idp/idpclient/client.go]: Done

File: idp/idpclient/client.go (hunk 6 above)
Comment (mtestrot) [reply idp/idpclient/client.go]: Done

**Pull request merged**